### DISCUSSION TITLE
Use spotless (maven) with java in vscode

### DISCUSSION THREAD

**Discussion by Theiaz** [diffplug/spotless#1763] [Q&A]
Hello,

my team and me are searching for a Java formatter. 

We've got an Java project which is built with maven. As IDEs Intellij IDEA and VSCode (with [Redhats Language Support package](https://marketplace.visualstudio.com/items?itemName=redhat.java) are used.

Spotless is working fine for Intellij IDEA. Unfortunally, it is not possible for me as an VSCode user to use spotless in an conveniet way. There is no plugin for maven and we do not want to migrate to gradle just because of the plugin support.

Some solutions which we may think of:
1. move from maven to gradle (plugin exists for IDEA and VSCode)
1. use other formatter
1. use pre-commit hook which applys `mvn spotless:apply`
1. use IDEA instead of vscode

As you can see every solution is not a real solution, but a workaround. I also thought about the possibility to integrate something like the [eclipse xml settings](https://raw.githubusercontent.com/google/styleguide/gh-pages/eclipse-java-google-style.xml) with spotless conventions (like described [here](https://code.visualstudio.com/docs/java/java-linting#_applying-formatter-settings)). However I am not able to find such a file and I don't think it even exists.

Are there any other solution for this kind of problem?

**Comment by nedtwigg** [comment]
The other solution would be to add an IDE hook into the maven plugin. [Here is the Gradle hook](https://github.com/diffplug/spotless/blob/main/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/IdeHook.java), only ~50 lines.

**Reply by SSmale** [reply]
I am also interested in this. Did you ever find an answer?

**Reply by Theiaz** [reply]
Nope, did not find any (easy installable) solution :(

**Reply by SSmale** [reply]
There is the run task on save that I might look at then and add it as a VSCode task

**Reply by SapiensAnatis** [reply]
Is there an issue for this? I'd like to try taking a look at a PR.

edit: ah, I see an IDE hook was already implemented for Maven in https://github.com/diffplug/spotless/pull/1782. So now someone just needs to write a VSCode extension. No promises but I might have a go at that instead :)

**Reply by SapiensAnatis** [reply]
Had a bit of a go over the past few evenings and got a proof-of-concept: https://github.com/SapiensAnatis/vscode-spotless-maven

It Works On My Machine (TM) but just does the dumbest thing possible and invokes `/usr/bin/mvn` whenever formatting is requested. So it probably won't work on Windows or any Linux machine that doesn't install `mvn` there. It also stops looking for `pom.xml` files as soon as it finds one, so will behave oddly if you have a mono-repo with multiple POMs. 

These constraints are largely because unlike the base Gradle for VSCode extension, the Maven extension doesn't appear to expose an API I can call to run targets, so I have to do all the project-finding and Maven invocation myself.

It's also really slow on account of invoking a mvn command for every format - about 1.5 seconds which is not a great experience particularly with formatting on save. The Gradle spotless extension sends requests to a Gradle server - and there is a server version of Maven apparently called `mvnd`, but it looks like it's tricky to install (not in any package manager repos) so not sure how to move forward with making the extension use that.

Because of the above, I wouldn't recommend it for general use right now. I'm aiming to continue working on it, but contributions/advice from anyone more experienced with Maven and/or VSCode extensions would be greatly appreciated :)

**Reply by nedtwigg** [reply]
Your project looks like a great start! Happy to merge any PRs which link to it from the Maven readme or anywhere else useful. Fine for it to be WIP, more eyeballs earlier means more helpers, hopefully!

**Reply by SapiensAnatis** [reply]
Thanks! I hit a bit of a roadblock because it doesn’t seem to work with mvnd - the IDE hook via stdin method just seems to hang when you use that. Need to troubleshoot that. I don’t think it’s really ready for general use until that major performance problem is solved so appreciate the offer about the links but will likely hold off.

**Reply by SapiensAnatis** [reply]
@nedtwigg I just heard back from a mvnd contributor that it doesn’t support goals that read from stdin :(

It seems like we have 3 choices:

1. Use regular mvn and accept the poor performance
2. Use mvnd and write to a temporary file somewhere for formatting
3. Maybe an alternative goal under regular mvn that starts a persistent gRPC or similar server? I think [CSharpier](https://github.com/belav/csharpier/tree/58ecd841ae814a6e9093f72e2abb6a915cb58b3a/Src/CSharpier.Cli/Server) does this for its IDE extensions. Not sure how feasible this is though or whether it’s something that belongs in Spotless

**Reply by nedtwigg** [reply]
I know nothing about mvnd, but mvnd + writing a file back and forth is my first instinct.

**Reply by SapiensAnatis** [reply]
Makes sense. I've been able to get a very hacky mvnd implementation working (and I mean [really hacky](https://github.com/SapiensAnatis/vscode-spotless-maven/blob/bd24fe3d5d655be20b064c22292a1a6d9dd87ef1/src/maven/daemonMavenExecutor.ts#L107)) using files for input and output, and even so it seems much more fluid and snappy - except for the first format when a daemon isn't ready yet.

Apparently, mvnd should support the basic stdin case we require, it just looks to have been broken at some point, so maybe we can ditch the files altogether soon! At which time the extension will be in a much better position.

**Reply by SapiensAnatis** [reply]
Quick visual comparison:

mvn:

![mvn](https://github.com/user-attachments/assets/5e7bf920-dc11-4230-928a-73fab7aa72be)

mvnd:

![mvnd](https://github.com/user-attachments/assets/382213ff-e8f4-4eab-abf0-13a284a2704b)

**Reply by bestbeforetoday** [reply]
@SapiensAnatis in case you didn't spot it, the release notes for mvnd [v2.0.0-rc-3](https://github.com/apache/maven-mvnd/releases/tag/2.0.0-rc-3) suggest that the stdin capability you wanted to use has been fixed.

**Reply by SapiensAnatis** [reply]
Thanks for the reminder @bestbeforetoday. I did have a go at building master with the commit that fixes it but didn't get much luck. Maybe I did something wrong though - I will try the pre-release.

The other issue is how mvnd (and regular mvn 4.0) crashes after running the IDE hook, so we have to ignore all errors :sweat_smile: hopefully that one gets fixed soon

**Comment by sashirestela** [comment]
A Spotless Maven extension for VSCode would be really great. I am interested in this.

**Comment by tkrisztian95** [comment]
As a workaround I'm using the VSCode plugin [Run On Save](https://marketplace.visualstudio.com/items?itemName=emeraldwalk.RunOnSave) to do run a `mvn spotless:apply` in the background whenever I hit save. 

Based on https://www.sethvargo.com/using-google-java-format-with-vs-code/ 

Slightly changed the command in the plugin settings in my `settings.json`. 
```
    "emeraldwalk.runonsave": {
        "commands": [
            {
              "match": "\\.java$",
              "cmd": "mvn spotless:apply"
            },
          ],
    }
```
This way it works with the configuration and stuff in the `pom.xml`.

Note: Unfortunately I cannot limit it to the single file with the `${file}`, so it always do apply Spotless on the whole project.

**Reply by RichieRock** [reply]
What you can do for a single file, is to install [google-java-format cli](https://github.com/google/google-java-format) globally on your machine and then use the same Run On Save plugin to run:

```
    "emeraldwalk.runonsave": {
        "commands": [
            {
                "match": "\\.java$",
                "cmd": "google-java-format --replace ${file}"
            },
        ],
    }
```

This way, it will format only the file you're editing and it won't run spotless apply for the whole project. 

The result should be the same if you have configured your spotless plugin to use google java format. For other formatting options, you'd need to pass the configuration to google-java-format cli.

**Reply by bestbeforetoday** [reply]
For a single file, you should be able to change the `cmd` to be `mvn spotless:apply -DspotlessFiles=${file}`

See here for details: https://github.com/diffplug/spotless/tree/main/plugin-maven#can-i-apply-spotless-to-specific-files

**Comment by okeylanders** [comment]
I fixed this by updating the pom to run the spotless apply goal before the check goal and to run during the compile phase. I also had to disable VSCode's own "format on save" setting ( to make sure they don't fight each other ).  I don't know if this is preferred or not, but it worked for me across multiple projects. 

Here are my steps:

1. Configure VS Code's  `settings.json` ( either User or Workspace ) to ensure that the Java files are not autoformatted. This prevents interference betwee the default formatter and spotless.

```json
"[java]": {
    "editor.formatOnSave": false
}
```

2. Update the maven spotless plugin goals and execution phase in the project pom. After updating, a "save" of any Java files covered by the maven lifecycle in VS Code will trigger the apply to run before the check and update any Java files. This will also prevent the constant errors from confusing the Language Service Provider.

```xml
<plugin>
    <groupId>com.diffplug.spotless</groupId>
    <artifactId>spotless-maven-plugin</artifactId>
    <configuration>
        <java>
            <palantirJavaFormat/>
        </java>
    </configuration>
    <executions>
        <execution>
            <goals>
                <goal>apply</goal>
                <goal>check</goal>
            </goals>
            <phase>compile</phase>
        </execution>
    </executions>
</plugin>
```
